### ISSUE TITLE
use batch read / writes, if available

### ISSUE THREAD

**Issue by marten-seemann** [lucas-clemente/quic-go#2607]
There's a [ReadBatch](https://godoc.org/golang.org/x/net/ipv4#PacketConn.ReadBatch) and a [WriteBatch]([WriteBatch](https://godoc.org/golang.org/x/net/ipv4#PacketConn.WriteBatch) on the `ipv4.PacketConn`. The equivalent functions are available on the `ipv6.PacketConn`.

We need to evaluate if we can either
* convert a `net.PacketConn` passed in by the the user to an `ipv4.PacketConn` or an `ipv6.PacketConn`, depending on the address, or otherwise
* ask users to pass in an `ipv{4,6}.PacketConn`, and then do an interface assertion, and use batched read / writes if possible

**Comment by marten-seemann** [comment]
Also see https://github.com/lucas-clemente/quic-go/issues/2586#issuecomment-642334615.

**Comment by bt90** [comment]
With https://github.com/lucas-clemente/quic-go/pull/3142 merged, what's the state of batch writes?

**Comment by marten-seemann** [comment]
> With #3142 merged, what's the state of batch writes?

It's not really clear if one would want to batch writes on a connection basis or aggregated over multiple (all?) connections. More benchmarking work would be needed to determine the best solution.

**Comment by bt90** [comment]
The main goal is to avoid syscall overhead IMHO. Go seems to use sendmmsg which allows different destinations.

**Comment by marten-seemann** [comment]
That's correct, it's just not clear to me which of the two options will give better performance.

**Comment by bt90** [comment]
Probably also depends on if go can use GSO for UDP and if that works with different destinations.

sendmmsg+GSO should be the goal as outlined by Cloudflares blog post:

https://blog.cloudflare.com/accelerating-udp-packet-transmission-for-quic/
